### ISSUE TITLE
Conversion Script 2 | ASCII to netCDF

### ISSUE THREAD

**Issue by dt-woods** [spatial-data-discovery/ay2021-1#9]
You have been given twelve (12) ASC raster data files containing monthly global Enhanced Vegetation Indexes (EVI) from the Terra satellite at 0.5-degree resolution (i.e., 720x360 pixels). An NGO has requested this data in a single netCDF file with the following requirements.

1.  netCDF file attributes need to include: "history," "contact," "institution," "title," and "satellite."
    - Use your contact information, William & Mary as the institution, the creation date as the history, "Monthly global Enhanced Vegetation Indexes (EVI) at 0.5 degree resolution" as the title and "Terra" as the satellite.
2.  EVI data should be saved as a single N-dimensional array with dimensions time, longitude, and latitude.
    - You must convert twelve individual data sets into a single array
3.  The netCDF file needs to include the variable "evi" with dimensions for "longitude," "latitude," and "time." It is also requested that variables be created for defining the values of "longitude," "latitude," and "time."
    -   For longitude and latitude, make sure your variables are in units of degrees_east and degrees_north (see [here](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#coordinate-types) for info on attributes) and that the coordinates are for pixel centroids (i.e., XLLCENTER and YLLCENTER). _You may have to calculate these._
    -   For time, use units should be "days since 1900-01-01" (note: check the literature to see how Python's `datetime` package defines time for easy calculations).
    -   EVI is a unitless index with a valid minimum of -0.2 and valid maximum of one; _note that the ASCII raster values have been scaled by a factor of 10000_ (see "Layers" info [here](https://lpdaac.usgs.gov/products/mod13c2v006/).
    -   Note, you should have three dimensions and four variables; please add an attribute to each variable describing its units (i.e., "deg_East," "deg_North," "days since 1900-01-01," and "unitless" for longitude, latitude, time, and EVI, respectively)
4.  EVI should have the appropriate fill value and missing value attributes, which should both be set to the NODATA_VALUE found in the ASCII raster header.
    - BONUS: include the correct scale factor attribute with the EVI variable (see more about scale factor attributes [here](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#variables))

Use what you know about ASC raster files and the information above and in the example script on GitHub to write a script that creates this netcdf file. Name the the script you wrote for the conversion `conv2_USERNAME.py` (no caps and use your GitHub username). Your script should read all the ASC raster files (found in zipped format in the /conv2/data folder) and produce a single .nc file in the classic format using scipy.io.netcdf. 

**Requirements:**

- Script finds and reads ASCII raster files from the given directory.
- Script parses the ASCII header from the data.
- Script creates an N-dimensional array from ASCII data.
- Script creates three dimensions (time, lat, lon), each with its own variable of the same name containing values in the appropriate units.
- Script creates a valid netcdf file (.nc) in the classic format.
- Script creates three dimensions and four variables requested by the client.
- Script assigns appropriate attributes to the file and variables.
- Script runs without errors.

**Grading**

| Score | Description |
| :--: | :--- |
| 0 | Did not submit |
| 2.5 | Does not meet the requirements of the assignment (code crashes, netcdf file is invalid/corrupt, missing attributes/variables/dimensions) |
| 3.5 | Almost everything works; maybe a typo or an error that's easily fixed |
| 4.5 | Meets all requirements |
| 5 | Meets and exceeds requirements (e.g., includes error handling, user feedback, the '-h' help message, etc.) |

Any scores less than full marks may be resubmitted using the within one week of grading tagging your commit message to the "Regrade" issue on GitHub.

**Comment by dt-woods** [comment]
I challenge you to make a drawing of what you think this looks like in a netCDF file. Post a photo of your drawings here.

**Comment by dt-woods** [comment]
Note that time units "days since 1900-01-01" auto converts to actual dates with Panoply ("days_since_1900" does not; note the spaces). See [here](http://www.bic.mni.mcgill.ca/users/sean/Docs/netcdf/guide.txn_18.html)

**Comment by dt-woods** [comment]
I am not as clear about the netCDF format itself, so I have some questions:

1) Is it correct to create dimensions "latitude" and "longitude" with values of 360 and 720, respectively?

> Yes, dimensions of `latitude` and `longitude` are just integer values representing the size of the dimension, which, in this case, is 360 and 720, respectively.
>

2) I'm trying to wrap my head around python datetime; why can't the time units be months, since it is monthly data? Is that part of the assignment, or a limitation I can overcome just by assigning values of 01,02,03...12 for months January until December to my time dimension?

> Yes, it is a part of the assignment. The client has requested that the values for the `time` variable be in units "days since 1900-01-01". Therefore, the variable `time` will be whole numbers (integers). The challenge is accounting for all the calendar day adjustments (e.g., leap years) between 1900 and 2012. This is where Python's `datetime` package comes in handy as it has a [timedelta](https://docs.python.org/3/library/datetime.html#datetime.timedelta) class that can return exactly what we are looking for.
> 

3) Closely following your video, would it be the correct interpretation that the RGB example is changed into months? This is to say that for the variable "EVI", I would have evi[0] = array for January, evi[1] = array for February... until evi[11] = array for December?

> Yes, that is a fair comparison. You may want to check out how [numpy array indexes](https://numpy.org/doc/stable/reference/arrays.indexing.html) work. Try this on your own is to see the shapes and values of different indexes with numpy arrays:

```py
import numpy
a = numpy.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])
a = a.reshape((2,4,3))
# What's the shape of a?
a.shape
# What's in each of these indexes and what is its shape: a[0], a[1]?
a[0].shape
```

**Comment by vincentandaw** [comment]
Professor @dt-woods, in what year and day are these monthly rasters taken?

**Comment by dt-woods** [comment]
It's in their file name, so you could scrub the filenames for `\d{4}-\d{2}-\d{2}` pattern (or just sub-string it because each filename is formatted the same way) to get the year, month and day, respectively.

**Comment by dt-woods** [comment]
Take a look at this function:

```py
import numpy
numpy.arange(-179.5, 180, 0.5)
```

**Comment by dt-woods** [comment]
My attempt at the drawing challenge.
![IMG_20201014_155601817](https://user-images.githubusercontent.com/29642997/96039040-1186c580-0e36-11eb-93c8-2097e7f28ebf.jpg)

**Comment by dt-woods** [comment]
Is your raster looking a bit upside down?

*Hint*: The [ASCII raster file](https://desktop.arcgis.com/en/arcmap/10.3/manage-data/raster-and-images/esri-ascii-raster-format.htm#GUID-CCB46F5D-098D-4C69-88D4-49F6D7C8CA81) is organized in such a way that the first row of data (below the header) is the top-most (or northern most) row of data.

Did you load your raster array with row 1 as the northern-most row but index your latitude variables beginning with the south-most values?

Try: 

* Flipping the order of your latitude values
* Flipping your row order of your 2D raster arrays (see [numpy.flipud](https://numpy.org/devdocs/reference/generated/numpy.flipud.html))

**Comment by sylviashea1** [comment]
Question: do we need to parse and save the heading information if we don't need to use it later? I hardcoded the cols, rows, lat and longs because they're the same for each file, so I don't see a use for keeping the header info (unless I'm missing something).

**Comment by dt-woods** [comment]
You're right. The headers in the Asc file are not really needed. You can hard code the values you need (e.g., NO DATA value) or just read them from the header once.
